contributors/design-proposals/scheduling/coscheduling.md

```diff
@@ -0,0 +1,174 @@
+    // will not start anyone.
+    NumMember int
+
+    // Resources defines the total resource the PodGroup requests to run
+    // Pods.
+    Resources v1.ResourceList
+}
+
+// PodGroupStatus represents the current state of a pod group.
+type PodGroupStatus struct {
+    // The number of actively running pods.
+    // +optional
+    Running int32
+
+    // The number of pods which reached phase Succeeded.
+    // +optional
+    Succeeded int32
+
+    // The number of pods which reached phase Failed.
+    // +optional
+    Failed int32
+}
+```
+
+The `PodGroup`, which is a namespaced object, specifies the behavior and status of a pod group, e.g. number of pods in a group. To define which pods are member of `PodGroup`, the following annotation key is introduced for `Pod`.
+
+```go
+scheduling.k8s.io/group-name
+```
+
+The `scheduling.k8s.io/group-name` annotation specifies the `PodGroup` that it belongs to; and the pod can only belong to the `PodGroup` in the same namespace. The pod, controlled by different collections, can also belong to the same `PodGroup`.  Because of performance concern,  it does not use `LabelSelector` to build the relationship between `PodGroup` and `Pod`.
+
+### Lifecycle Management
+
+As the lifecycle of Pods in PodGroup maybe different from controller to controller, the lifecycle management is delegated to the owner of Pods. In kube-arbitrator, it'll record related events for controller to manage pods, e.g. `Unschedulable`.
+
+The update to `PodGroup` is not support for now; and deleting `PodGroup` does not impact Pod's status.
+
+### kube-arbitrator
+
+`kube-arbitrator` only watches `PodGroup` and `Pod`. It'll reconstruct 'Job' by annotation of Pod and `PodGroup`, the `Pod`s are considered as 'Task' of 'Job'; if annotation is empty, `kube-arbitrator` records an unschedulable event of pod to ask user/controller to resubmit it to `kube-scheduler`. `kube-arbitrator` does not schedule pods until its `PodGroup` is ready.
+
+As `kube-arbitrator` and `kube-scheduler` maybe running in parallel; `kube-arbitrator` follows multi-scheduler feature to only handle the `Pod` that submitted to `kube-arbitrator`. The `kube-arbitrator` does scheduling as follow:
+
+1. Reconstruct 'Job' by the annotation of `Pod` and `PodGroup`
+2. If `PodGroup` is not ready, the 'job' will not be scheduled; and an unschedulable event of `Pod` will be recorded
+3. In `allocate` phase, `kube-arbitrator` will
+   * record an `Unschedulable` event of `PodGroup` if some pods are running but `succeeded + pending + running < minAvailable`, the `PodGroupController` takes action according to `LifecyclePolicy`
+   * allocate (but not bind) resource to Pods according to Pod's spec, e.g. `NodeAffinity`
+   * bind all Pods to hosts until job ready (`numMember` <= `allocated Pods` + `succeeded Pods`)
+4. If can not allocate enough resources to the job, the pods are keeping pending; and the resource can not allocate to other job
```

Comment:
s/pods are keeping pending/pods stay pending/
s/resource can not allocate/resources cannot be allocated/